src/analysis/calculus/cont_diff.lean

```diff
@@ -2264,3 +2265,308 @@ lemma cont_diff.restrict_scalars (h : cont_diff 𝕜' n f) :
+      ‖iterated_fderiv_within 𝕜 n (λ (y : Du), B.precompL Du (fderiv_within 𝕜 f s y) (g y)) s x‖
+          ≤ ‖B.precompL Du‖ * ∑ (i : ℕ) in finset.range (n + 1), n.choose i
+            * ‖iterated_fderiv_within 𝕜 i (fderiv_within 𝕜 f s) s x‖
+            * ‖iterated_fderiv_within 𝕜 (n - i) g s x‖ :
+        IH _ (hf.fderiv_within hs In) (hg.of_le (nat.cast_le.2 (nat.le_succ n)))
+      ... ≤ ‖B‖ * ∑ (i : ℕ) in finset.range (n + 1), n.choose i
+            * ‖iterated_fderiv_within 𝕜 i (fderiv_within 𝕜 f s) s x‖
+            * ‖iterated_fderiv_within 𝕜 (n - i) g s x‖ :
+        mul_le_mul_of_nonneg_right (B.norm_precompL_le Du) (finset.sum_nonneg' (λ i, by positivity))
+      ... = _ :
+        begin
+          congr' 1,
+          apply finset.sum_congr rfl (λ i hi, _ ),
+          rw [iterated_fderiv_within_succ_eq_comp_right hs hx, linear_isometry_equiv.norm_map],
+        end,
+    have J : iterated_fderiv_within 𝕜 n
+      (λ (y : Du), fderiv_within 𝕜 (λ (y : Du), B (f y) (g y)) s y) s x
+      = iterated_fderiv_within 𝕜 n (λ y, B.precompR Du (f y) (fderiv_within 𝕜 g s y)
+        + B.precompL Du (fderiv_within 𝕜 f s y) (g y)) s x,
+    { apply iterated_fderiv_within_congr hs (λ y hy, _) hx,
+      have L : (1 : with_top ℕ) ≤ n.succ,
+        by simpa only [enat.coe_one, nat.one_le_cast] using nat.succ_pos n,
+      exact B.fderiv_within_of_bilinear (hf.differentiable_on L y hy)
+        (hg.differentiable_on L y hy) (hs y hy) },
+    rw [iterated_fderiv_within_succ_eq_comp_right hs hx, linear_isometry_equiv.norm_map, J],
+    have A : cont_diff_on 𝕜 n (λ y, B.precompR Du (f y) (fderiv_within 𝕜 g s y)) s,
+      from (B.precompR Du).is_bounded_bilinear_map.cont_diff.comp_cont_diff_on₂
+        (hf.of_le (nat.cast_le.2 (nat.le_succ n))) (hg.fderiv_within hs In),
+    have A' : cont_diff_on 𝕜 n (λ y, B.precompL Du (fderiv_within 𝕜 f s y) (g y)) s,
+      from (B.precompL Du).is_bounded_bilinear_map.cont_diff.comp_cont_diff_on₂
+        (hf.fderiv_within hs In) (hg.of_le (nat.cast_le.2 (nat.le_succ n))),
+    rw iterated_fderiv_within_add_apply' A A' hs hx,
+    apply (norm_add_le _ _).trans ((add_le_add I1 I2).trans (le_of_eq _)),
+    rw ← mul_add,
+    congr' 1,
+    exact (sum_choose_succ_mul_mul_sub (λ i, ‖iterated_fderiv_within 𝕜 i f s x‖)
+      (λ i, ‖iterated_fderiv_within 𝕜 i g s x‖) n).symm }
+end
+
+/-- Bounding the norm of the iterated derivative of `B (f x) (g x)` within a set in terms of the
+iterated derivatives of `f` and `g` when `B` is bilinear:
+`‖D^n (x ↦ B (f x) (g x))‖ ≤ ‖B‖ ∑_{k ≤ n} n.choose k ‖D^k f‖ ‖D^{n-k} g‖` -/
+lemma continuous_linear_map.norm_iterated_fderiv_within_le_of_bilinear
+  (B : E →L[𝕜] F →L[𝕜] G) {f : D → E} {g : D → F} {N : with_top ℕ} {s : set D} {x : D}
+  (hf : cont_diff_on 𝕜 N f s) (hg : cont_diff_on 𝕜 N g s) (hs : unique_diff_on 𝕜 s) (hx : x ∈ s)
+  {n : ℕ} (hn : (n : with_top ℕ) ≤ N) :
+  ‖iterated_fderiv_within 𝕜 n (λ y, B (f y) (g y)) s x‖
```

Comment:
Would it make sense to split all the `iterated_fderiv_within` lemmas (including the existing ones) into a new file? At a glance, none of the `cont_diff` lemmas use `iterated_fderiv_within` in their proofs, and this file is rather long already.

Reply:
I had already tried that, and it doesn't work well since it fact `iterated_fderiv` is used crucially in some of the proofs on `cont_diff` (notably to relate `cont_diff` and `cont_diff_on_univ`).